lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3038,7 +3038,8 @@ bool LLVMToSPIRVBase::transDecoration(Value *V, SPIRVValue *BV) {
     if (Opcode == Instruction::FAdd || Opcode == Instruction::FSub ||
         Opcode == Instruction::FMul || Opcode == Instruction::FDiv ||
         Opcode == Instruction::FRem ||
-        ((Opcode == Instruction::FNeg || Opcode == Instruction::FCmp) &&
+        ((Opcode == Instruction::FNeg || Opcode == Instruction::FCmp ||
```

Comment:
@MrSidims, could you check if SPIRVReader handles fast-math flags for FNeg and FCmp correctly, please?
If not, please, file a tracker to follow-up.

Unfortunately, SPIR-V format is not able to translate LLVM semantics correctly. @bashbaug, SPIR-V doesn't allow applying fast-math flags to OpSelect and OpCall instructions, but LLVM IR can have fast math flags attached to `select` and `call` instructions. Do you know if this is going to be addressed by future SPIR-V versions?

Reply:
> SPIR-V doesn't allow applying fast-math flags to OpSelect and OpCall instructions, but LLVM IR can have fast math flags attached to select and call instructions

This is a part of [SPV_KHR_float_controls2](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_float_controls2.html):

> Replace the text following "Only valid on …​" with:
> All core instructions which use any floating-point type for either operands or result.
> OpExInst extended instructions, where expressly permitted by the extended instruction set in use.

It may be a bit too much and some instructions can or should be excluded from the list, but it is what it is.

Reply:
Given:

> OpExInst extended instructions, where expressly permitted by the extended instruction set in use.

Do we need some updates to the OpenCL Extended Instruction set, or perhaps the OpenCL SPIR-V environment spec?  Feel free to file a SPIR-V or OpenCL issue for tracking, if so, so this doesn't slip through the cracks.

We should also add some targeted CTS tests for this functionality, even if they aren't exhaustive.

Reply:
> could you check if SPIRVReader handles fast-math flags for FNeg and FCmp correctly, please

@bader according to passing tests https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/test/transcoding/fcmp.ll and https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/test/transcoding/fneg.ll it handles correctly. Manual run also doesn't show anything fishy.

Reply:
Thanks!

Reply:
> Do we need some updates to the OpenCL Extended Instruction set, or perhaps the OpenCL SPIR-V environment spec?

@bashbaug no, https://registry.khronos.org/SPIR-V/specs/unified1/OpenCL.ExtendedInstructionSet.100.pdf already contains the required lines like:
```
For environments that allow use of FPFastMathMode decorations on OpExtInst instructions,
FPFastMathMode decorations may be applied to the math instructions.
```
and there are other several references about it.
But I believe Alexey had another question related to `select` and `call` instructions.

> We should also add some targeted CTS tests for this functionality, even if they aren't exhaustive.

We should, https://github.com/KhronosGroup/OpenCL-CTS/pulls?q=is%3Apr+is%3Aopen+FPFastMathMode search shows nothing